File: ffw_bringup/ffw_bringup/joint_trajectory_splitter.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from trajectory_msgs.msg import JointTrajectory
from trajectory_msgs.msg import JointTrajectoryPoint


def _select_joints(message, indices):
    selected = JointTrajectory()
    selected.header = message.header
    selected.joint_names = [message.joint_names[index] for index in indices]

    for point in message.points:
        selected_point = JointTrajectoryPoint()
        for field in ('positions', 'velocities', 'accelerations', 'effort'):
            values = getattr(point, field)
            setattr(
                selected_point,
                field,
                [values[index] for index in indices] if values else [],
            )
        selected_point.time_from_start = point.time_from_start
        selected.points.append(selected_point)

    return selected


def split_trajectory(message):
    joint_count = len(message.joint_names)
    if not joint_count:
        raise ValueError('trajectory has no joint names')

    for point in message.points:
        for field in ('positions', 'velocities', 'accelerations', 'effort'):
            values = getattr(point, field)
            if values and len(values) != joint_count:
                raise ValueError(
                    f'{field} has {len(values)} values for {joint_count} joints'
                )

    arm_indices = []
    gripper_indices = []
    for index, name in enumerate(message.joint_names):
        if name.startswith('gripper_'):
            gripper_indices.append(index)
        else:
            arm_indices.append(index)
    return (
        _select_joints(message, arm_indices),
        _select_joints(message, gripper_indices),
    )


# Split /leader/joint_trajectory_command_broadcaster_{left,right}/joint_trajectory
# into modular arm and gripper controller topics.
class JointTrajectorySplitter(Node):

    def __init__(self):
        super().__init__('joint_trajectory_splitter')

        self._trajectory_publishers = {}
        for side, suffix in (('left', 'l'), ('right', 'r')):
            self._trajectory_publishers[side] = (
                self.create_publisher(
                    JointTrajectory, f'/arm_{suffix}_controller/joint_trajectory', 10
                ),
                self.create_publisher(
                    JointTrajectory, f'/gripper_{suffix}_controller/joint_trajectory', 10
                ),
            )
            self.create_subscription(
                JointTrajectory,
                f'/leader/joint_trajectory_command_broadcaster_{side}/joint_trajectory',
                lambda message, side=side: self._split(side, message),
                10,
            )

    def _split(self, side, message):
        try:
            arm, gripper = split_trajectory(message)
        except ValueError as error:
            self.get_logger().warning(f'Ignored invalid {side} trajectory: {error}')
            return

        arm_publisher, gripper_publisher = self._trajectory_publishers[side]
        if arm.joint_names:
            arm_publisher.publish(arm)
        if gripper.joint_names:
            gripper_publisher.publish(gripper)


def main(args=None):
    rclpy.init(args=args)
    node = JointTrajectorySplitter()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == '__main__':
    main()

File: ffw_joint_trajectory_command_broadcaster/src/joint_trajectory_splitter.cpp
#include <memory>
#include <stdexcept>
#include <utility>
#include <string>
#include <vector>

#include "rclcpp/rclcpp.hpp"
#include "trajectory_msgs/msg/joint_trajectory.hpp"

namespace ffw_trajectory_splitter
{
using Trajectory = trajectory_msgs::msg::JointTrajectory;
using Point = trajectory_msgs::msg::JointTrajectoryPoint;

static std::pair<Trajectory, Trajectory> split_trajectory(const Trajectory & message)
{
  if (message.joint_names.empty()) {
    throw std::invalid_argument("trajectory has no joint names");
  }

  const auto fields = {&Point::positions, &Point::velocities, &Point::accelerations,
    &Point::effort};
  for (const auto & point : message.points) {
    for (const auto field : fields) {
      const auto & values = point.*field;
      if (!values.empty() && values.size() != message.joint_names.size()) {
        throw std::invalid_argument("trajectory field length does not match joint names");
      }
    }
  }

  Trajectory arm, gripper;
  for (auto * output : {&arm, &gripper}) {
    output->header = message.header;
    output->points.resize(message.points.size());
    for (size_t p = 0; p < message.points.size(); ++p) {
      output->points[p].time_from_start = message.points[p].time_from_start;
    }
  }
  for (size_t i = 0; i < message.joint_names.size(); ++i) {
    const auto & name = message.joint_names[i];
    auto & output = name.rfind("gripper_", 0) == 0 ? gripper : arm;
    output.joint_names.push_back(name);
    for (size_t p = 0; p < message.points.size(); ++p) {
      for (const auto field : fields) {
        const auto & values = message.points[p].*field;
        if (!values.empty()) {
          (output.points[p].*field).push_back(values[i]);
        }
      }
    }
  }
  return {std::move(arm), std::move(gripper)};
}

class JointTrajectorySplitter : public rclcpp::Node
{
public:
  JointTrajectorySplitter()
  : Node("joint_trajectory_splitter")
  {
    for (const std::string side : {"left", "right"}) {
      const std::string suffix = side == "left" ? "l" : "r";
      auto arm = create_publisher<Trajectory>(
        "/arm_" + suffix + "_controller/joint_trajectory", 10);
      auto gripper = create_publisher<Trajectory>(
        "/gripper_" + suffix + "_controller/joint_trajectory", 10);
      subscriptions_.push_back(create_subscription<Trajectory>(
        "/leader/joint_trajectory_command_broadcaster_" + side + "/joint_trajectory", 10,
        [this, side, arm, gripper](Trajectory::ConstSharedPtr message) {
          try {
            auto outputs = split_trajectory(*message);
            if (!outputs.first.joint_names.empty()) {
              arm->publish(outputs.first);
            }
            if (!outputs.second.joint_names.empty()) {
              gripper->publish(outputs.second);
            }
          } catch (const std::invalid_argument & error) {
            RCLCPP_WARN(
              get_logger(), "Ignored invalid %s trajectory: %s", side.c_str(), error.what());
          }
        }));
    }
  }

private:
  std::vector<rclcpp::Subscription<Trajectory>::SharedPtr> subscriptions_;
};
}  // namespace ffw_trajectory_splitter

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<ffw_trajectory_splitter::JointTrajectorySplitter>());
  rclcpp::shutdown();
  return 0;
}
